### PR TITLE
Wallet: create account using recovery phrase

### LINKED ISSUE #19514: Wallet: Create Account - Import Using Recovery Phrase
Implement creating an account using recovery phrase.

Designs: https://www.figma.com/file/HKncH4wnDwZDAhc4AryK8Y/Wallet-for-Mobile?type=design&node-id=2070-592351&mode=dev

### PR DESCRIPTION
**Pull request by OmarBasem**
fixes: https://github.com/status-im/status-mobile/issues/19514

### Summary: 
This PR implements recovering a keypair using recovery phrase and creating a wallet account from that keypair.

The `enter-seed-phrase` screen inside onboarding moved it to a common namespace to be used for both of wallet and onboarding recovery.

---

[Designs](https://www.figma.com/file/HKncH4wnDwZDAhc4AryK8Y/Wallet-for-Mobile?type=design&node-id=2070-592351&mode=dev)

For QA testing, areas that may be impacted:
- Onboarding: recovery phrase

### Demo:

https://github.com/status-im/status-mobile/assets/29354102/c00c8b8d-dace-415b-a562-1cbe1b7a32c7

### CODE REVIEW

File: src/status_im/common/enter_seed_phrase/view.cljs
Comment (OmarBasem) [inline src/status_im/common/enter_seed_phrase/view.cljs:1]: Moved `enter-seed-phrase` to be in a common namespace

File: src/status_im/contexts/wallet/add_account/create_account/select_keypair/view.cljs
Comment (OmarBasem) [inline src/status_im/contexts/wallet/add_account/create_account/select_keypair/view.cljs]: Not used in MVP

**Review by briansztamfater — APPROVED** [review]
LGTM

File: src/status_im/common/enter_seed_phrase/view.cljs
Comment (J-Son89) [inline src/status_im/common/enter_seed_phrase/view.cljs:163]: 🙌

File: src/status_im/common/enter_seed_phrase/view.cljs
Comment (J-Son89) [inline src/status_im/common/enter_seed_phrase/view.cljs]: keypair? is a big ambiguous imo. What does this imply/ is it used for?

File: src/status_im/contexts/wallet/add_account/create_account/events.cljs
Comment (J-Son89) [inline src/status_im/contexts/wallet/add_account/create_account/events.cljs]: why do we refer to it as `secret-phrase` here by `seed-phrase` elsewhere? should it not be consitent?

File: src/status_im/contexts/wallet/add_account/create_account/events_test.cljs
Comment (J-Son89) [inline src/status_im/contexts/wallet/add_account/create_account/events_test.cljs]: 🙏

File: src/status_im/contexts/wallet/add_account/create_account/new_keypair/check_your_backup/view.cljs
Comment (J-Son89) [inline src/status_im/contexts/wallet/add_account/create_account/new_keypair/check_your_backup/view.cljs]: hmm, not a fan of overriding vars like this. Can we avoid it? 
Is there a general consensus on doing things like this in cljs? 
@ilmotta wdyt? 🤔

File: src/status_im/contexts/wallet/effects.cljs
Comment (J-Son89) [inline src/status_im/contexts/wallet/effects.cljs]: should we make this safer? 🤔 
i.e if it's a string do this, if it's a vector use string/join and otherwise return an empty string/nil? 🤔

File: src/status_im/navigation/screens.cljs
Comment (J-Son89) [inline src/status_im/navigation/screens.cljs]: it's a nit, but should we not have these consistent?
i.e 
` `` 
{:name      :screen/wallet.enter-seed-phrase
     :component enter-seed-phrase/view}
 ```
 or 
 ```
 {:name :screen/wallet.recovery-phrase
 :component recovery-phrase/view}

File: translations/en.json
Comment (J-Son89) [inline translations/en.json:2549]: please check this with design, as I think it might have been a design bug to add the space :)

**Review by J-Son89 — APPROVED** [review]
some minor details that would be good to address. 

Otherwise looks great, nice one @OmarBasem! 🚀

File: src/status_im/common/enter_seed_phrase/view.cljs
Comment (OmarBasem) [reply src/status_im/common/enter_seed_phrase/view.cljs]: Means are we recovering a keypair ?

Changed it to `recovering-keypair?`

File: src/status_im/contexts/wallet/add_account/create_account/new_keypair/check_your_backup/view.cljs
Comment (OmarBasem) [reply src/status_im/contexts/wallet/add_account/create_account/new_keypair/check_your_backup/view.cljs]: Changed the second var to `unmasked-secret-phrase`

File: src/status_im/contexts/wallet/effects.cljs
Comment (OmarBasem) [reply src/status_im/contexts/wallet/effects.cljs]: @J-Son89 But we shouldn't get to that point with an empty string. It will be of no use and the status-go call will fail. So probably better to make it fail earlier?

File: translations/en.json
Comment (OmarBasem) [reply translations/en.json:2549]: Checked with design team, it should "Key pair"

File: src/status_im/contexts/wallet/effects.cljs
Comment (J-Son89) [reply src/status_im/contexts/wallet/effects.cljs]: tbh, I'm not sure. Just flaggin this to be careful. Whatever you think is best! 🙏

File: src/status_im/contexts/wallet/add_account/create_account/new_keypair/check_your_backup/view.cljs
Comment (ilmotta) [reply src/status_im/contexts/wallet/add_account/create_account/new_keypair/check_your_backup/view.cljs]: @J-Son89, @OmarBasem I did my fair share of shadowing let bindings in Clojure in the name of convenience. The consensus in the team is to not shadow, not even local bindings. AFAIR this is not in our guidelines.

Currently, clj-kondo can only protect us from shadowed vars and shadowed function args, but not shadowed local bindings.

**Comment by VolodLytvynenko** [comment]
Hi @OmarBasem . Thank you for PR. Take a look found issues:
## PR_ISSUE 1: Users are able to add an account using an invalid seed phrase when 18 or 24 random valid words are entered

### Steps:
1. Go to 'Use recovery phrase' page
2. Enter 18 or 24 random valid words (example: `general general general general general general general general general general general general general general general general general general general general general general general general`)


### Actual result:
The account is successfully added, even though the seed phrase is invalid.

### Expected result:
Potential solution:
- Implement validation for seed phrases entered when adding an account 
image

Let's see what design team can reply
https://discord.com/channels/624634427930312714/852533718790570015/1231952024321396776

### OS: 
IOS, Android
### Devices:
- Pixel 7a, Android 13
- iPhone 11 Pro Max, IOS 17

**Comment by VolodLytvynenko** [comment]
## PR_ISSUE 2: The key pair field requires 6 words instead of 5 and the validation message "Key pair name must be at least 5 characters" is not displayed in case if less than 5 chars entered

### Steps:
1. Go to key pair screen
2. Enter 5 valid chars


### Actual result:
- Key pair creation can't be confirmed if key pair name with 5 valid characters is eneted.
- The validation message "Key pair name must be at least 5 characters" is not displayed when less than 5 characters are entered.
image

### Expected result:
- Key pair should be able to be created with 5 valid chars 
- The validation message "Key pair name must be at least 5 characters" is displayed when less than 5 characters are entered.https://www.figma.com/file/HKncH4wnDwZDAhc4AryK8Y/Wallet-for-Mobile?node-id=10820%3A259608&mode=dev

### OS: 
IOS, Android

### Devices:
- Pixel 7a, Android 13
- iPhone 11 Pro Max, IOS 17

**Comment by VolodLytvynenko** [comment]
## PR_ISSUE 3: The color of the  'continue' button does not inherit the color of the account

### Steps:
1. Set up a particular color picker for the account
2. Go to 'Use recovery phrase' page 

### Actual result:
The 'continue' button does not inherit the color of the account
image

### Expected result:
The 'continue' button  inherits the color of the account

### OS: 
IOS, Android

### Devices:
- Pixel 7a, Android 13
- iPhone 11 Pro Max, IOS 17

**Comment by VolodLytvynenko** [comment]
## Question: 
Should the derivation work correctly in the case of account importing, or is this something that will only work after [this](https://github.com/status-im/status-mobile/issues/19472) issue will be fixed? Currently, all imported accounts have an invalid derivation path

**Comment by OmarBasem** [comment]
HI @VolodLytvynenko, thanks for your review!

Issue 1 is fixed.

Issue 2: fixed the problem requiring 6 characters instead of 5,

> "Key pair name must be at least 5 characters" is not displayed in case if less than 5 chars entered

This would require to display this red error before the user even starts typing or as soon as the user starts typing which I think would look odd. I think keeping the button disabled until 5 characters have been typed gets the message across. (I will confirm with design team also)

Issue 3: It is correct as it is. It should use the profile color (also similar to other buttons in keypair flow)

**Comment by VolodLytvynenko** [comment]
Hi @OmarBasem, thank you for your reply. Could you please check https://github.com/status-im/status-mobile/pull/19702#issuecomment-2069684608 if this is not an issue?

**Comment by OmarBasem** [comment]
> Hi @OmarBasem, thank you for your reply. Could you please check [#19702 (comment)](https://github.com/status-im/status-mobile/pull/19702#issuecomment-2069684608) if this is not an issue?

@VolodLytvynenko sorry missed the question. The derivation path will work correctly after that issue is fixed.

**Comment by VolodLytvynenko** [comment]
Hi @OmarBasem 
> Issue 3: It is correct as it is. It should use the profile color (also similar to other buttons in keypair flow)

I didn't understand you here. Is it correct the buttons are always blue, regardless of the account color that is set up?

> @VolodLytvynenko sorry missed the question. The derivation path will work correctly after that issue is fixed.

Thank you for the clarification!

**Comment by VolodLytvynenko** [comment]
@OmarBasem  PR is tested and ready to be merged. Still, some follow-ups are required, which I will create after approval from the design team [here](https://discord.com/channels/624634427930312714/852533718790570015/1231952024321396776)

**Comment by OmarBasem** [comment]
Thanks for your testing @VolodLytvynenko. 

> I didn't understand you here. Is it correct the buttons are always blue, regardless of the account color that is set up?

Yes correct the button color when creating an account depends on the profile color - not the account yet to be created.

**Pull request merged**
